lib/valueptr.h

```diff
@@ -47,6 +50,10 @@ class CPPCHECKLIB ValuePtr {
     ValuePtr(const U& value) : mPtr(cloner<U>::apply(&value)), mClone(&cloner<U>::apply)
     {}
 
+    template<class U>
+    ValuePtr(U&& value) : mPtr(cloner<U>::move(&value)), mClone(&cloner<U>::apply)
```

Comment:
We should only move when its passed as an rvalue references, not for lvalue references. Also, this might be called for the copy constructor and move constructor, so it needs to be constrained.

```cpp
    template<class U, REQUIRES("Must be rvalue", !is_lvalue_reference<U>), REQUIRES("Must not be ValuePtr", !std::is_base_of<ValuePtr, U>)>
    ValuePtr(U&& value) : mPtr(cloner<U>::move(&value)), mClone(&cloner<U>::apply)
    {}
```

Reply:
Thanks for the input. I need to check again what the actual impact of this was - probably not within the next few days.

I just pushed it again to clean up my local branches.

Reply:
One case where I came across this was `makeAnalyzer`. You would expect that it is being moved on return but it is copied so we essentially create it twice every time - with the destruction on top.